src/main/java/org/casbin/jcasbin/main/SyncedEnforcer.java

```diff
@@ -95,22 +96,819 @@ public SyncedEnforcer(String modelPath, String policyFile, boolean enableLog) {
         super(modelPath, policyFile, enableLog);
     }
 
+    /**
+     * setWatcher sets the current watcher.
+     *
+     * @param watcher the watcher.
+     */
+    public void setWatcher(Watcher watcher) {
+        this.watcher = watcher;
+        watcher.setUpdateCallback(this::loadPolicy);
+    }
+
+    /**
+     * clearPolicy clears all policy.
+     */
+    public void clearPolicy() {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            super.clearPolicy();
+        } finally {
+            READ_WRITE_LOCK.writeLock().unlock();
+        }
+    }
+
+    /**
+     * loadPolicy reloads the policy from file/database.
+     */
+    public void loadPolicy() {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            super.loadPolicy();
+        } finally {
+            READ_WRITE_LOCK.writeLock().unlock();
+        }
+    }
+
+    /**
+     * loadFilteredPolicy reloads a filtered policy from file/database.
+     *
+     * @param filter the filter used to specify which type of policy should be loaded.
+     */
+    public void loadFilteredPolicy(Object filter) {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
+            super.loadFilteredPolicy(filter);
+        } finally {
+            READ_WRITE_LOCK.writeLock().unlock();
+        }
+    }
+
+    /**
+     * savePolicy saves the current policy (usually after changed with
+     * Casbin API) back to file/database.
+     */
+    public void savePolicy() {
+        try {
+            READ_WRITE_LOCK.readLock().lock();
+            super.savePolicy();
+        } finally {
+            READ_WRITE_LOCK.readLock().unlock();
+        }
+    }
+
+    /**
+     * buildRoleLinks manually rebuild the
+     * role inheritance relations.
+     */
+    public void buildRoleLinks() {
+        try {
+            READ_WRITE_LOCK.readLock().lock();
+            super.buildRoleLinks();
+        } finally {
+            READ_WRITE_LOCK.readLock().unlock();
+        }
+    }
+
+    /**
+     * enforce decides whether a "subject" can access a "object" with
+     * the operation "action", input parameters are usually: (sub, obj, act).
+     *
+     * @param rvals the request needs to be mediated, usually an array
+     *              of strings, can be class instances if ABAC is used.
+     * @return whether to allow the request.
+     */
+    public boolean enforce(Object... rvals) {
+        try {
+            READ_WRITE_LOCK.writeLock().lock();
```

Comment:
Since there is no write operation, I think it's better to use `readLock`.

Reply:
@hsluoyz Same question in Golang.

Reply:
Is Golang wrong here?

Reply:
I think writeLock should be preferable here as while processing enforcement there should be no deletion or addition in config

Reply:
While read must be allowed while on-going enforcement processing

Reply:
When use the `readLock` , It would check for `WriteLock`.